File: server/models/order.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict
from datetime import datetime

class OrderItem(BaseModel):
    name: Optional[str]
    qty: Optional[int]
    price: Optional[float]
    product: Optional[str]  # Reference to Product ID
    image: Optional[str]
    variant_image: Optional[str]
    selected_variants: Optional[Dict]
    selected_size: Optional[str]
    selected_color: Optional[str]
    color_code: Optional[str]
    variant: Optional[str]
    sku: Optional[str]
    variant_id: Optional[str]

class ContactDetails(BaseModel):
    first_name: Optional[str]
    last_name: Optional[str]
    email: Optional[str]
    phone: Optional[str]

class Address(BaseModel):
    address: Optional[str]
    apartment: Optional[str]
    city: Optional[str]
    state_region: Optional[str]
    postal_code: Optional[str]
    country: Optional[str]

class Order(BaseModel):
    order_id: str
    user: Optional[str]  # Reference to User ID
    guest_id: Optional[str]  # UUID for guest tracking
    order_items: List[OrderItem] = []
    contact_details: ContactDetails
    shipping_address: Address
    billing_address: Address
    payment_method: Optional[str]
    payment_result: Optional[Dict]
    items_price: Optional[float]
    tax_price: Optional[float]
    shipping_price: Optional[float]
    total_price: Optional[float]
    is_paid: bool = False
    paid_at: Optional[datetime]
    is_delivered: bool = False
    delivered_at: Optional[datetime]
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

File: server/models/Order.js
const mongoose = require('mongoose');

const orderItemSchema = new mongoose.Schema({
  name: String,
  qty: Number,
  price: Number,
  product: { type: mongoose.Schema.Types.ObjectId, ref: 'Product' },
  image: String,
  variantImage: { type: String }, // Image of the selected variant
  selectedVariants: { type: Object },
  selectedSize: { type: String },
  selectedColor: { type: String },
  colorCode: { type: String }, // Hex color code for visual display
  variant: { type: String }, // Variant name/type
  sku: { type: String }, // Product SKU
  variantId: { type: String },
});

const orderSchema = new mongoose.Schema({
  orderId: { type: String, unique: true, index: true, required: true },
  user: { type: mongoose.Schema.Types.ObjectId, ref: 'User', sparse: true }, // Allow null for guest orders
  guestId: { type: String, sparse: true, index: true }, // UUID for guest tracking
  orderItems: [orderItemSchema],
  contactDetails: {
    firstName: String,
    lastName: String,
    email: String,
    phone: String,
  },
  shippingAddress: {
    address: String,
    apartment: String,
    city: String,
    stateRegion: String,
    postalCode: String,
    country: String,
  },
  billingAddress: {
    address: String,
    apartment: String,
    city: String,
    stateRegion: String,
    postalCode: String,
    country: String,
  },
  paymentMethod: String,
  paymentResult: { type: Object },
  itemsPrice: Number,
  taxPrice: Number,
  shippingPrice: Number,
  totalPrice: Number,
  couponCode: { type: String, default: null }, // Applied coupon code
  discountAmount: { type: Number, default: 0 }, // Discount amount applied

  paidAt: Date,
  fulfillmentStatus: { type: String, enum: ['unfulfilled', 'fulfilled'], default: 'unfulfilled' }, // Order fulfillment status
  deliveryStatus: { type: String, enum: ['', 'shipped', 'delivered', 'refunded'], default: '' }, // Delivery status field
  deliveredAt: Date,
  status: { type: String, default: 'pending' }, // Payment/processing status
  remarks: { type: String, default: '' },
}, { timestamps: true });

const Order = mongoose.model('Order', orderSchema);
module.exports = Order;

